Issue #8 - Expand the context concept

uthentic-ai opened the pull request:
This pull request implements the changes requested in Issue #8. It expands the context concept to include both contact and thread contexts, adds getter and setter methods, records timestamps for context entries, and implements both in-memory and PostgreSQL storage engines.

Closes #8

TelegramSam commented on context.py:
```diff
@@ -0,0 +1,61 @@
+        if key in contexts:
+            del contexts[key]
+
+class Context:
```
Unify the getting and setting functions so that the default is thread, and an optional argument will cause it to be set or loaded from the contact. The get function should default to thread, but will also return the value from the contact if it exists. Getting a value returns an object that contains the value, the source (context or thread), and the timestamp it was set.

TelegramSam commented on router.py:
```diff
@@ -49,41 +48,52 @@ def register_handler(self, from_did, msg_type, handler):
             handler = self.handler_map.get(handler_name)
             if handler:
-                await self.scheduler.spawn(handler(msg, context))
+                await self.scheduler.spawn(handler(msg, contact_context, thread_context))
```
only one context object - choosing contact or thread as a destination is handled by the get and set methods as described in my other comment.

uthentic-ai commented:
I've addressed the comments and made the following changes:

1. Modified the Context class in context.py:
   - Created a unified 'get' method that defaults to thread context but also checks contact context.
   - Created a unified 'set' method with an optional argument to specify contact or thread context.
   - Updated the return value of 'get' to include the value, source, and timestamp.

2. Updated the MessageRouter class in router.py:
   - Now using a single context object instead of separate contact and thread contexts.
   - Updated the route_message method to use the new unified context methods.

These changes should address the concerns raised in the review. Please let me know if any further modifications are needed.
